<a href="https://colab.research.google.com/github/esnue/ThesisAllocationSystem/blob/main/topic_model.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
'''
from https://github.com/MaartenGr/BERTopic
'''

In [1]:
!pip install bertopic[all]

     |████████████████████████████████| 81kB 5.7MB/s 
     |████████████████████████████████| 6.4MB 16.3MB/s 
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done
     |████████████████████████████████| 450kB 35.3MB/s 
     |████████████████████████████████| 13.2MB 249kB/s 
     |████████████████████████████████| 2.0MB 57.5MB/s 
     |████████████████████████████████| 1.2MB 51.7MB/s 
     |████████████████████████████████| 983kB 50.2MB/s 
     |████████████████████████████████| 19.7MB 168kB/s 
     |████████████████████████████████| 798kB 56.5MB/s 
     |████████████████████████████████| 71kB 9.5MB/s 
     |████████████████████████████████| 890kB 49.0MB/s 
     |████████████████████████████████| 3.2MB 50.7MB/s 
  Created wheel for hdbscan: filename=hdbscan-0.8.27-cp37-cp37m-linux_x86_64.whl size=2311677 sha256=771186a894e3aae5a70871b4ac16d7ecb6cf3e7ffe5ccadb5e7be07a58b29e66
  Stored in directory: /root/.cache/pip/

# **Prepare Data**

We use the .csv that contains the supervisors' publicly available papers as training data. 

In [12]:
from bertopic import BERTopic
import pandas as pd

#use csv file containing all supervisor papers, currently: broken file
train_df = pd.read_csv('/content/drive/MyDrive/ThesisAllocationSystem/train-papers-txt/train-papers-txt.csv', encoding ='latin1')
print(train_df.head())
content = train_df["Content"]

print(f"Shape of dataframe is {content.shape}")

             FileName                                            Content
0   Hammerschmid3.txt  The Governance of Infrastructure    Edited by ...
1   Hammerschmid6.txt  PUBLIC PRIVATE PARTNERSHIP BETWEEN EUPHORIA AN...
2  Hammerschmid10.txt  More delegation, more political control? Polit...
3   Hammerschmid1.txt  1   Curry, D., Hammerschmid, G., Jilke, S., Va...
4   Hammerschmid2.txt  2007 EGPA_paper1109.doc   See discussions, sta...
Shape of dataframe is (411,)


In [5]:
print(content.head())


0    The Governance of Infrastructure    Edited by ...
1    PUBLIC PRIVATE PARTNERSHIP BETWEEN EUPHORIA AN...
2    More delegation, more political control? Polit...
3    1   Curry, D., Hammerschmid, G., Jilke, S., Va...
4    2007 EGPA_paper1109.doc   See discussions, sta...
Name: Content, dtype: object


# **Create Topics**

We select the "multilingual" as the main language for our training data documents, as parts of the text might be in languages other than English, depending on the publishing organization's location.

In [22]:
topic_model = BERTopic(language="multilingual", calculate_probabilities=True)
topics, probs = topic_model.fit_transform(content)

In [23]:
topic_model.get_topic_freq().head(5)

Topic  Count
0     -1    109
1      2     57
2      3     44
3      5     41
4      1     36

-1 refers to all outliers and should typically be ignored. Next, let's take a look at the most frequent topic that was generated, `topic 2`:

In [24]:
topic_model.get_topic(2)

[('german', 0.016236009665040393),
 ('are', 0.01479233685282645),
 ('germany', 0.01341021740016272),
 ('from', 0.01284794428661778),
 ('be', 0.011950023178241677),
 ('education', 0.011063593609595478),
 ('vet', 0.011041236341394657),
 ('training', 0.011023451014940294),
 ('not', 0.010182324285342543),
 ('have', 0.010010263605742628)]

In [25]:
print(topics)
print(len(topics))

[-1, -1, 5, 5, -1, 5, -1, 5, 5, -1, 5, 5, -1, 6, -1, 5, 5, 5, -1, 2, 2, 2, 2, 2, -1, 2, -1, 2, 2, 2, 2, 2, -1, -1, -1, 6, -1, -1, 2, 1, 7, -1, -1, 7, -1, -1, -1, 7, 7, -1, -1, -1, -1, -1, -1, 7, 5, 7, -1, 1, 4, -1, -1, 7, -1, 7, -1, -1, -1, 6, 1, -1, 7, 1, 6, 4, -1, 7, 8, 8, -1, -1, -1, 8, 8, -1, -1, -1, 8, 8, -1, 8, 8, 3, 8, 8, 2, 8, -1, 8, -1, 8, 8, 2, -1, 8, 8, 8, 8, 8, 8, 8, -1, -1, 8, 8, -1, 8, 8, 3, -1, 8, 3, 3, 3, 3, 3, 3, 3, 3, 3, -1, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 8, 3, -1, 3, 3, 3, 3, -1, 3, 3, 3, -1, 3, -1, 3, 3, -1, 3, 3, 3, 3, -1, 3, 3, 3, 3, 3, 3, 3, -1, 1, 1, 1, 1, 1, 1, -1, 1, 1, 1, 1, 1, 1, -1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 4, 4, -1, 4, 4, 4, 4, 4, 5, 4, 4, 5, -1, 4, 4, 5, 4, -1, 4, -1, 4, -1, 4, 4, 4, 4, 4, 4, 4, 4, 4, -1, -1, -1, 2, 2, 2, -1, 2, -1, -1, -1, 5, 2, 2, 2, -1, -1, 2, 2, -1, 2, -1, 2, 2, 2, 2, -1, 2, 6, 6, 2, 6, -1, -1, 6, 6, 6, 6, -1, 6, 5, -1, 6, -1, 6, 2, 6, 5, 6, 4, -1, 4, 6, 2, 6, -1, 2, -1, 2, 5, 2, 5, 2, 5, 2, -1, -1, 2, 4, 2

# **Visualize Topics**

Visualize generated topics.

In [26]:
topic_model.visualize_topics()

## **Label Training Data**

We can use the BERTopic generated topics to label our training data. 

In [28]:
train_df["topic"] = topics

print(train_df.head())

             FileName                                            Content  topic
0   Hammerschmid3.txt  The Governance of Infrastructure    Edited by ...     -1
1   Hammerschmid6.txt  PUBLIC PRIVATE PARTNERSHIP BETWEEN EUPHORIA AN...     -1
2  Hammerschmid10.txt  More delegation, more political control? Polit...      5
3   Hammerschmid1.txt  1   Curry, D., Hammerschmid, G., Jilke, S., Va...      5
4   Hammerschmid2.txt  2007 EGPA_paper1109.doc   See discussions, sta...     -1


As mentioned before, when `topic` is equal to -1, this refers to an outlier topic. Let's take a closer look.

In [29]:
print(train_df.loc[train_df['topic'] == -1])

              FileName  ... topic
0    Hammerschmid3.txt  ...    -1
1    Hammerschmid6.txt  ...    -1
4    Hammerschmid2.txt  ...    -1
6    Hammerschmid9.txt  ...    -1
9    Hammerschmid8.txt  ...    -1
..                 ...  ...   ...
393         mena12.txt  ...    -1
395         mena14.txt  ...    -1
398  jachtenfuchs4.txt  ...    -1
399         mena15.txt  ...    -1
400         mena16.txt  ...    -1

[109 rows x 3 columns]


109 rows are classified as outliers, probably because it is a very specialized research question. However, the number of text classified as outliers is a non-trivial share of the overall training data, hence we won't drop them. 